### DISCUSSION TITLE
Is there a best practice to structure projects

### DISCUSSION THREAD

**Discussion by Keith-CY** [nervosnetwork/capsule#124] [Q&A]
I've learned from @zhengjianhui that most projects maintained by capsule have a `libs`-like module outside the `contracts` module for regulatory reasons.

- kuai-mvp-demo: https://github.com/ckb-js/kuai/tree/feat/mvp-contract/packages/samples/mvp-dapp/contract/types
- das-contracts: https://github.com/dotbitHQ/das-contracts/tree/master/libs

Is there an explicit guide for this convention?

I ask this because [kuai](https://github.com/ckb-js/kuai) has a project template including the contracts module, and if the `libs`-like module is advisable for contracts development, it will be included in the project template.

**Comment by blckngm** [comment]
I would put the libs in a `crates` or `libs` dir alongside `contracts` and add them to the root cargo workspace.

**Reply by Keith-CY** [reply]
Will this be a convention and clarified in capsule's development guide? If so, it will be common knowledge of contract developers.

**Reply by Keith-CY** [reply]
I would suggest mentioning it in capsule's document so all teams will following the same pattern, which is also good for code analysis
